sdk-java/src/main/java/ly/count/sdk/java/internal/ModuleViews.java

```diff
@@ -1,9 +1,571 @@
+        currentViewID = currentViewData.viewID;
+
+        Map<String, Object> viewSegmentation = createViewEventSegmentation(currentViewData, firstView, true, customViewSegmentation);
+
+        if (firstView) {
+            L.d("[ModuleViews] Recording view as the first one in the session. [" + viewName + "]");
+            firstView = false;
+        }
+
+        recordView(currentViewID, 0.0, viewSegmentation);
+        return currentViewData.viewID;
+    }
+
+    void stopViewWithNameInternal(@Nullable String viewName, @Nullable Map<String, Object> customViewSegmentation) {
+        String viewID = validateViewWithName(viewName, "stopViewWithNameInternal");
+        if (viewID == null) {
+            return;
+        }
+
+        stopViewWithIDInternal(viewID, customViewSegmentation);
+    }
+
+    void stopViewWithIDInternal(@Nullable String viewID, @Nullable Map<String, Object> customViewSegmentation) {
+        ViewData vd = validateViewID(viewID, "stopViewWithIDInternal");
+        if (vd == null) {
+            return;
+        }
+        removeReservedKeysFromViewSegmentation(customViewSegmentation);
+
+        L.d("[ModuleViews] View [" + vd.viewName + "], id:[" + vd.viewID + "] is getting closed, reporting duration: [" + (TimeUtils.uniqueTimestampS() - vd.viewStartTimeSeconds) + "] s, current timestamp: [" + TimeUtils.uniqueTimestampMs() + "]");
+        recordViewEndEvent(vd, customViewSegmentation, "stopViewWithIDInternal");
+
+        viewDataMap.remove(vd.viewID);
+    }
+
+    void recordView(String id, Double duration, Map<String, Object> segmentation) {
+        ModuleEvents events = internalConfig.sdk.module(ModuleEvents.class);
+        if (events == null) {
+            L.e("[ModuleViews] recordView, events module is not initialized");
+            return;
+        }
+
+        events.recordEventInternal(KEY_VIEW_EVENT, 1, 0.0, duration, segmentation, id);
+    }
+
+    void recordViewEndEvent(ViewData vd, @Nullable Map<String, Object> filteredCustomViewSegmentation, String viewRecordingSource) {
+        long lastElapsedDurationSeconds = 0;
+        //we do sanity check the time component and print error in case of problem
+        if (vd.viewStartTimeSeconds < 0) {
+            L.e("[ModuleViews] " + viewRecordingSource + ", view start time value is not normal: [" + vd.viewStartTimeSeconds + "], ignoring that duration");
+        } else if (vd.viewStartTimeSeconds == 0) {
+            L.i("[ModuleViews] " + viewRecordingSource + ", view is either paused or didn't run, ignoring start timestamp");
+        } else {
+            lastElapsedDurationSeconds = TimeUtils.uniqueTimestampS() - vd.viewStartTimeSeconds;
+        }
+
+        //only record view if the view name is not null
+        if (vd.viewName == null) {
+            L.e("[ModuleViews] stopViewWithIDInternal, view has no internal name, ignoring it");
+            return;
+        }
+
+        long viewDurationSeconds = lastElapsedDurationSeconds;
```

Comment:
i think 'viewDurationSeconds' is not needed, we can use directly 'lastElapsedDurationSeconds'